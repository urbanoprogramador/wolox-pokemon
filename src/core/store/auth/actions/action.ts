import { asyncTodos, IUser, typeAuth } from "../reducers/reducer";
import { asyncMac, mac } from "../../utils/configReducer";


import { signInWithPopup, GoogleAuthProvider, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from "firebase/auth";
import { auth } from "../../../firebase/firebase-config";

const [
  actionAuthPending,
  actionAuthError,
  actionAuthSuccess
] = asyncMac(asyncTodos);

interface IActionLogin{
  payload:IUser
};


export const actionAuthUser = mac<IActionLogin>(typeAuth.login /* , 'payload' */ );
export const actionAuthLogout = mac<null>(typeAuth.logout);



export const actionClearAuth = () => (dispatch:Function) => {
  dispatch(actionAuthLogout(null));
  localStorage.removeItem('user');
}

export const asyncStartGoogleLogin = () => (dispatch:Function) => {
  dispatch(actionAuthPending());
  const provider = new GoogleAuthProvider();
  signInWithPopup(auth, provider)
    .then((result) => {
      const user = (result.user as any);

      dispatch(actionAuthUser({
        payload:{
        displayName: user.displayName,
        email: user.email,
        token: user.accessToken
        }
      }));
      localStorage.setItem('user', JSON.stringify({
        displayName: user.displayName,
        email: user.email,
        token: user.accessToken,
      }));
      dispatch(actionAuthSuccess());


      // ...
    }).catch((error) => {

      //auth/popup-closed-by-user
      if (error.code === "auth/popup-closed-by-user") {
        dispatch(actionAuthError("ventana cerrada por el usuario "));
      } else {
        dispatch(actionAuthError("Error Desconocido "));
      }

    });
}

export const asyncCreateUserWithEmailAndPassword = ({ email, password, displayname }:{email:string,password:string,displayname:string}) => (dispatch:Function) => {
  dispatch(actionAuthPending());

  createUserWithEmailAndPassword(auth, email, password)
    .then((userCredential) => {
      // Signed in 
      auth.currentUser &&
        updateProfile(auth.currentUser, {
          displayName: displayname
        }).then(() => {
          // Profile updated!
          // ...
          const user = (userCredential.user as any);
          dispatch(actionAuthUser({
            payload:{
              displayName: user.displayName,
              email: user.email,
              token: user.accessToken,
            }
          }));
          localStorage.setItem('user', JSON.stringify({
            displayName: user.displayName,
            email: user.email,
            token: user.accessToken,
          }));
          dispatch(actionAuthSuccess());
        }).catch((error) => {
          console.log(error);
        });

      // ...
    })
    .catch((error) => {

      if (error.code === "auth/email-already-in-use" || error.code === "auth/user-not-found") {
        dispatch(actionAuthError("Usuario ya se encuentra registrado"));
      } else {
        dispatch(actionAuthError("Error Desconocido "));
      }
    });

}


export const asyncLoginEmailAndPassword = ({ email, password }:{email:string,password:string}) => (dispatch:Function) => {
  dispatch(actionAuthPending());
  signInWithEmailAndPassword(auth, email, password)
    .then((userCredential) => {
      // Signed in
      const user = (userCredential.user as any);
      dispatch(actionAuthUser({
        payload:{
          displayName: user.displayName,
            email: user.email,
            token: user.accessToken,
        }
      }));
      localStorage.setItem('user', JSON.stringify({
        displayName: user.displayName,
        email: user.email,
        token: user.accessToken,
      }));
      dispatch(actionAuthSuccess());
      // ...
    })
    .catch((error) => {

      /* 
      "auth/wrong-password"
      auth/user-not-found
      */

      if (error.code === "auth/wrong-password" || error.code === "auth/user-not-found") {
        dispatch(actionAuthError("Usuario o contraseÃ±a invalido"));
      } else {
        dispatch(actionAuthError("Error Desconocido "));
      }
    });
}

export const asyncSignOut = () => (dispatch:Function) => {
  dispatch(actionAuthPending());
  signOut(auth).then(() => {
    dispatch(actionClearAuth());
    dispatch(actionAuthSuccess());
  }).catch((error) => {
    dispatch(actionAuthError("Error Desconocido "));
  });
}