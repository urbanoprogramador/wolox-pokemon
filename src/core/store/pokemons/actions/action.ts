import { asyncPokemon, typePokemon } from "../reducers/reducer";
import { asyncMac, mac } from "../../utils/configReducer";
import { IDetallePokemon, IResponsePokemon } from "../pokemonInterface";
import { IActionPokemonEntities, IPokemon } from './../pokemonInterface';

const [
  actionPokemonPending,
  actionPokemonError,
  actionPokemonSuccess
] = asyncMac(asyncPokemon);







export const actionLoadProduct = mac<IActionPokemonEntities>(typePokemon.load);
export const actionLoadCompare = mac<IActionPokemonEntities>(typePokemon.loadCompare);
export const actionLoadFavorite = mac<IActionPokemonEntities>(typePokemon.loadFavorite);





export const asyncLoadPokemon = () => (dispatch: Function) => {

  dispatch(actionPokemonPending());
  fetch('https://pokeapi.co/api/v2/pokemon/?offset=0&limit=5000')
    .then(res => res.json())
    .then((d: IResponsePokemon) => {
      const p = d.results.map<IPokemon>(({ name, url }): IPokemon => {
        const id = url.split('/')[6];
        return {
          name: name,
          id
        }
      });
      dispatch(actionLoadProduct({
        payload: p
      }));
      dispatch(actionPokemonSuccess());
      localStorage.setItem('pokemon', JSON.stringify(p));

    }).catch((error) => {
      dispatch(actionPokemonError({ payload: error.message }));
    }).finally(() => {

    });

}

export const asyncLoadPokemonDetailPokemon = (pokemon: IPokemon) => (dispatch: Function, getState: Function) => {

  const allPokemon = (getState().pokemon.entities) as IPokemon[];

  dispatch(actionPokemonPending());
  fetch('https://pokeapi.co/api/v2/pokemon/' + pokemon.id)
    .then(res => res.json())
    .then((d: IDetallePokemon) => {

      const temporal: IPokemon = {
        ...pokemon,
        detalle: d
      };

      const temporalEnties = allPokemon.map((todo) => {
        if (todo.id === temporal.id) {
          return temporal;
        }
        return todo;
      });

      localStorage.setItem('pokemon', JSON.stringify(temporalEnties));

      dispatch(actionLoadProduct({
        payload: temporalEnties
      }));

    }).catch((error) => {
      //dispatch(actionPokemonError({ payload: error.message }));
    }).finally(() => {

    });



}




export const asyncComparePokemon = (pokemon: IPokemon,isCompare:boolean) => (dispatch: Function, getState: Function) => {
  
  const allPokemon = (getState().pokemon.entities) as IPokemon[];
  const compare = (getState().pokemon.compare) as IPokemon[];
  const tCompare = [ ...compare ];

  if(isCompare){
    dispatch(actionLoadCompare({
      payload: tCompare.filter((e)=>e.id!==pokemon.id)
    }));
    return ;
  }

  

  if (compare.length >= 3) {
    tCompare.shift();
  }
  if (pokemon.detalle) {
    
    tCompare.push(pokemon);

    

    dispatch(actionLoadCompare({
      payload: tCompare
    }));
  } else {
    fetch('https://pokeapi.co/api/v2/pokemon/' + pokemon.id)
      .then(res => res.json())
      .then((d: IDetallePokemon) => {

        const temporal: IPokemon = {
          ...pokemon,
          detalle: d
        };

        const temporalEnties = allPokemon.map((todo) => {
          if (todo.id === temporal.id) {
            return temporal;
          }
          return todo;
        });

        localStorage.setItem('pokemon', JSON.stringify(temporalEnties));

        dispatch(actionLoadProduct({
          payload: temporalEnties
        }));

        tCompare.push(temporal);

        dispatch(actionLoadCompare({
          payload: tCompare
        }));

      }).catch((error) => {
        //dispatch(actionPokemonError({ payload: error.message }));
      }).finally(() => {

      });
  }


}








export const asyncLoadFavorite = (pokemon: IPokemon,isFavorite:boolean) => (dispatch: Function, getState: Function) => {
  
  const allPokemon = (getState().pokemon.entities) as IPokemon[];
  const favorite = (getState().pokemon.favorite) as IPokemon[];
  const tFavorite = [ ...favorite ];

  if(isFavorite){
    dispatch(actionLoadFavorite({
      payload: tFavorite.filter((e)=>e.id!==pokemon.id)
    }));
    return ;
  }

  if (pokemon.detalle) {
    
    tFavorite.push(pokemon);
    dispatch(actionLoadFavorite({
      payload: tFavorite
    }));
  } else {
    fetch('https://pokeapi.co/api/v2/pokemon/' + pokemon.id)
      .then(res => res.json())
      .then((d: IDetallePokemon) => {

        const temporal: IPokemon = {
          ...pokemon,
          detalle: d
        };

        const temporalEnties = allPokemon.map((todo) => {
          if (todo.id === temporal.id) {
            return temporal;
          }
          return todo;
        });

        localStorage.setItem('pokemon', JSON.stringify(temporalEnties));

        dispatch(actionLoadProduct({
          payload: temporalEnties
        }));

        tFavorite.push(temporal);

        dispatch(actionLoadFavorite({
          payload: tFavorite
        }));

      }).catch((error) => {
        //dispatch(actionPokemonError({ payload: error.message }));
      }).finally(() => {

      });
  }


}