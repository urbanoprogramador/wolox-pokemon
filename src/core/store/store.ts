import { middlewareReduxInpect} from './middleware/middleware';
import { combineReducers, createStore } from 'redux';
import { userReducer } from './auth/reducers/reducer';
import { themeReducer } from './theme/reducers/reducer';
import { pokemonReducer } from './pokemons/reducers/reducer';


const reducer=combineReducers({
  auth:userReducer,
  pokemon:pokemonReducer,
  theme:themeReducer
})

export const store = createStore(reducer,middlewareReduxInpect);