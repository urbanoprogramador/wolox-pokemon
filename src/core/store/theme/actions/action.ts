import { typeTheme } from "../reducers/reducer";
import {  mac } from "../../utils/configReducer";
import { IActionConfigTheme } from "../themeInterface";


export const actionLoadConfigTheme = mac<IActionConfigTheme>(typeTheme.loadConfig);

