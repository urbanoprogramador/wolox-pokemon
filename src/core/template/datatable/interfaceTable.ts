import { Dispatch, SetStateAction } from "react";


export interface ItemplateHtml{
    (obj:object)
}

export interface IFieldname{
    fieldname:string,
    key?:string,
    html?:ItemplateHtml
}

export interface IDatatableContext {
    data:object[];
    filterData: object[]
    setFilterData: Dispatch<SetStateAction<object[]>>
    sizes: number[]
    setSizes: Dispatch<SetStateAction<number[]>>
    page: number
    setPage:  Dispatch<SetStateAction<number>>
    fieldname:IFieldname[],
    size:number,
    setSize:Dispatch<SetStateAction<number>>

}