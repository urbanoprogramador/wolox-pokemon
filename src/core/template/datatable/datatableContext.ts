import { createContext } from "react";
import { IDatatableContext } from "./interfaceTable";

//context
export const DatatableContext = createContext<IDatatableContext>({} as IDatatableContext);

