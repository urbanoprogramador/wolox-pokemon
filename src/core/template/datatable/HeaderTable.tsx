import React, { useContext } from "react";
import { DatatableContext } from './datatableContext';


export const HeaderTable = () => {

    const {fieldname}=useContext(DatatableContext);
     
 


    return (

    <thead>
        <tr>
        {
            fieldname.map(({fieldname})=>{
                return (<th key={fieldname}>
                {fieldname}
                </th>)
            })
        }
        </tr>
    </thead>

    );
}