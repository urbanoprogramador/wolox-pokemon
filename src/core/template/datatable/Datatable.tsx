
import React from 'react';
import { DatatableContext } from './datatableContext';
import { useState } from 'react';
import { IFieldname } from './interfaceTable';
import { HeaderTable } from './HeaderTable';
import { TableBody } from './TableBody';
import { TableActions } from './TableActions';
import { TablePaginate } from './TablePaginate';

type Props<T> = {
    data: T[],
    fieldname: IFieldname[],
    sizes?: number[]
}


export const Datatable: React.FC<Props<{}>> = ({ data, fieldname, sizes: sizesT = [5, 10, 20, 50, 100] }) => {

    const [filterData, setFilterData] = useState<object[]>([...data]);
    const [sizes, setSizes] = useState<number[]>(sizesT);
    const [page, setPage] = useState<number>(0);

    const [size, setSize] = useState(20);

    return (<DatatableContext.Provider value={{
        data,
        filterData,
        setFilterData,
        sizes,
        setSizes,
        page,
        setPage,
        fieldname,
        size,
        setSize
    }}>
        <div className='dataTableContent'>
            <TablePaginate />
            <TableActions />
            <table className='dataTable'>
                <HeaderTable />
                <TableBody />
            </table>

            <TablePaginate />
        </div>

    </DatatableContext.Provider>);
}