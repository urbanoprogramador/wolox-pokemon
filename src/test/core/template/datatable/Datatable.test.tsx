import React from 'react';
import { fireEvent, render, waitFor } from "@testing-library/react";
jest.setTimeout(10000);

import '@testing-library/jest-dom'
import { Datatable } from '../../../../core/template/datatable/Datatable';
import { Provider } from 'react-redux';
import { store } from './../../../../core/store/store';
import { actionLoadConfigTheme } from 'core/store/theme/actions/action';
import { data } from './data';


describe("Probando el DataTable", () => {

  test('Carga completa de el componente table y sus partes', () => {

    store.dispatch(actionLoadConfigTheme({ payload: { width: 700 } }));
    render(<Provider store={store}>
      < Datatable data={data} fieldname={[

        {
          fieldname: 'id',
          key: 'id'

        }, {
          fieldname: 'nombre',
          key: 'nombre'

        },
        {
          fieldname: 'Correo',
          key: 'email'
        },
        {
          fieldname: 'telefono',
          key: 'telefono',
        }
      ]} />
    </Provider>);


    expect(document.querySelector('.dataTableContent')).toMatchSnapshot();
    //uexpect(screen.debug()).toBe('uno');
  });


  test('Prueba de que se pasa un parametro custom para el header de la tabla ', () => {

    store.dispatch(actionLoadConfigTheme({ payload: { width: 700 } }));
    render(<Provider store={store}>
      < Datatable data={data} fieldname={[

        {
          fieldname: 'id',
          key: 'id'

        }, {
          fieldname: 'nombre',
          key: 'nombre'

        },
        {
          fieldname: 'Correo',
          key: 'email'
        },
        {
          fieldname: 'telefono',
          key: 'telefono',
        }
      ]} />
    </Provider>);

    expect(document.querySelectorAll('.dataTableContent th')[2].innerHTML).toBe('Correo');
    expect(document.querySelectorAll('.dataTableContent th')[3].innerHTML).toBe('telefono');
    //uexpect(screen.debug()).toBe('uno');
  });
  test('Prueba a la input busqueda del datatable buscando por "usuario uno"', async () => {

    store.dispatch(actionLoadConfigTheme({ payload: { width: 700 } }));
    render(<Provider store={store}>
      < Datatable data={data} fieldname={[

        {
          fieldname: 'id',
          key: 'id'

        }, {
          fieldname: 'nombre',
          key: 'nombre'

        },
        {
          fieldname: 'Correo',
          key: 'email'
        },
        {
          fieldname: 'telefono',
          key: 'telefono',
        }
      ]} />
    </Provider>);
    const input = document.querySelector('input');
    if (input) {
      fireEvent.change(input, { target: { value: 'usuario uno' } });
    }
    await waitFor(() => {expect(document.querySelectorAll('tbody td')[2].innerHTML).toEqual("correo@usuario.uno"); }, {
      timeout: 5000,
    });

    //expect(document.querySelectorAll('tbody').length).toBe(1);
    //uexpect(screen.debug()).toBe('uno');
  });


  test('Prueba a la input busqueda del datatable buscando algo que no existe "abcdefjs"', async () => {

    store.dispatch(actionLoadConfigTheme({ payload: { width: 700 } }));
    render(<Provider store={store}>
      < Datatable data={data} fieldname={[

        {
          fieldname: 'id',
          key: 'id'

        }, {
          fieldname: 'nombre',
          key: 'nombre'

        },
        {
          fieldname: 'Correo',
          key: 'email'
        },
        {
          fieldname: 'telefono',
          key: 'telefono',
        }
      ]} />
    </Provider>);
    const input = document.querySelector('input');
    if (input) {
      fireEvent.change(input, { target: { value: 'abcdefjsasdfasdf' } });
    }
    await waitFor(() => {expect(document.querySelector('tbody td')?.innerHTML).toEqual("Item No encontrado"); }, {
      timeout: 5000,
    });

    
    //uexpect(screen.debug()).toBe('uno');
  });

});
