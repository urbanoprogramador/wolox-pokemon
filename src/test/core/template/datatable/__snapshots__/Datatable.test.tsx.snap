// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el DataTable Carga completa de el componente table y sus partes 1`] = `
<div
  class="dataTableContent"
>
  <div
    class="contenPagi"
  >
    <div>
      <h4>
        Pokemones 
        1
         a 
        5
         de  
        5
          
      </h4>
      <ul
        class="pagination"
      >
        <li>
          <i
            aria-hidden="true"
            class="fa fa-fast-backward"
          />
        </li>
        <li>
          <i
            aria-hidden="true"
            class="fa fa-backward"
          />
        </li>
        <li
          class="active"
        >
          1
        </li>
        <li>
          <i
            aria-hidden="true"
            class="fa fa-forward"
          />
        </li>
        <li>
          <i
            aria-hidden="true"
            class="fa fa-fast-forward"
          />
        </li>
      </ul>
    </div>
  </div>
  <div>
    <div>
      <select>
        <option
          value="5"
        >
          5
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="20"
        >
          20
        </option>
        <option
          value="50"
        >
          50
        </option>
        <option
          value="100"
        >
          100
        </option>
      </select>
      <input
        type="text"
        value=""
      />
    </div>
  </div>
  <table
    class="dataTable"
  >
    <thead>
      <tr>
        <th>
          id
        </th>
        <th>
          nombre
        </th>
        <th>
          Correo
        </th>
        <th>
          telefono
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          usuario uno
        </td>
        <td>
          correo@usuario.uno
        </td>
        <td>
          12344444444
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          usuario dos
        </td>
        <td>
          correo@usuario.dos
        </td>
        <td>
          22222222
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          usuario tres
        </td>
        <td>
          correo@usuario.tres
        </td>
        <td>
          33333333
        </td>
      </tr>
      <tr>
        <td>
          4
        </td>
        <td>
          usuario cuatro
        </td>
        <td>
          correo@usuario.cuatro
        </td>
        <td>
          444444
        </td>
      </tr>
      <tr>
        <td>
          5
        </td>
        <td>
          usuario cinco
        </td>
        <td>
          correo@usuario.cionco
        </td>
        <td>
          555555555
        </td>
      </tr>
    </tbody>
  </table>
  <div
    class="contenPagi"
  >
    <div>
      <h4>
        Pokemones 
        1
         a 
        5
         de  
        5
          
      </h4>
      <ul
        class="pagination"
      >
        <li>
          <i
            aria-hidden="true"
            class="fa fa-fast-backward"
          />
        </li>
        <li>
          <i
            aria-hidden="true"
            class="fa fa-backward"
          />
        </li>
        <li
          class="active"
        >
          1
        </li>
        <li>
          <i
            aria-hidden="true"
            class="fa fa-forward"
          />
        </li>
        <li>
          <i
            aria-hidden="true"
            class="fa fa-fast-forward"
          />
        </li>
      </ul>
    </div>
  </div>
</div>
`;
